### ISSUE TITLE
Conflict with NETStandard.Library package being downgraded

### ISSUE THREAD

**Issue by mrahhal** [autofac/Autofac.Extensions.DependencyInjection#25]
The dependencies of this package causes conflicts:

```
MyApp -> Autofac.Extensions.DependencyInjection 4.2.0 -> Microsoft.Extensions.DependencyInjection.Abstractions 1.1.0 -> NETStandard.Library (>= 1.6.1) 

MyApp -> Autofac.Extensions.DependencyInjection 4.2.0 -> NETStandard.Library (>= 1.6.0)
```

Is it possible to release an updated version that depends on version 2 of `Microsoft.Extensions.DependencyInjection.Abstractions`?

**Comment by tillig** [comment]
Can you show a small repro .csproj with this happening? We don't upgrade dependencies unless there's a demonstrated need we can verify and going to 2.0 on the abstractions library is a breaking change for people still on ASP.NET Core 1.x.

**Comment by mrahhal** [comment]
You're right about that being a breaking change, I really meant releasing a new major version.

But anyway, forgetting about that for a second. At the very least, a dependency on NETStandard.Library 1.6.1 can be taken, instead of 1.6.0. If you look at the dependencies of [Microsoft.Extensions.DependencyInjection.Abstractions 1.1.0](https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection.Abstractions/1.1.0), it's fairly obvious that it depends on 1.6.1, while this package depends no 1.6.0.

This is what causes this conflict.

**Comment by mrahhal** [comment]
I think the reason why this conflict didn't appear in your own usage is because you're referencing NETStandard.Library 1.6.1 (or 1.6.0) in your app project. This breaks the conflict and explicitly tells the resolver which version to use, making it happy.

In our case, we target the full .net framework so we don't need a NETStandard.Library reference, that's when this conflict appears.

**Comment by tillig** [comment]
Again, if you can show us a minimal repro project that illustrates the issue, we can look into it. Simply describing it is not enough; we have several projects, from examples in other repos to work projects, where this doesn't happen. We need to literally see a .csproj that we can build and witness this happening.

**Comment by mrahhal** [comment]
Will do a minimal repro soon. (Just wanted to explain why it's happening)

**Comment by mrahhal** [comment]
[You can find the repro here](https://github.com/mrahhal/AutofacDIRepro).

Repro steps:
- Clone
- dotnet restore

```
$ dotnet restore
  Restoring packages for C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.csproj...
C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.csproj : error NU1605: Detected package downgrade: NETStandard.Library from 1.6.1 to 1.6.0. Reference the package directly from the project to select a different version.  [C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.sln]
C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.csproj : error NU1605:  AutofacDIRepro -> Autofac.Extensions.DependencyInjection 4.2.0 -> Microsoft.Extensions.DependencyInjection.Abstractions 1.1.0 -> NETStandard.Library (>= 1.6.1)  [C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.sln]
C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.csproj : error NU1605:  AutofacDIRepro -> Autofac.Extensions.DependencyInjection 4.2.0 -> NETStandard.Library (>= 1.6.0) [C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.sln]
  Restore failed in 293.29 ms for C:\D\dev\src\Projects\AutofacDIRepro\AutofacDIRepro.csproj.
```

As you can see. The simplest project, a console app targeting full .net that just simply references `Autofac.Extensions.DependencyInjection` and nothing else experiences this problem.

Workaround would be to explicitly reference either Microsoft.Extensions.DependencyInjection.Abstractions 2.0.0, or NETStandard.Library 1.6.1, even though I don't depend on either. This breaks the conflict.

Of course, our real world use case is much more complex, with multiple N-Tier levels of projects that depend on each other, some of which are apps, and most of which are just libraries.

**Comment by tillig** [comment]
Thanks, we'll take a look.

Note the "workaround" you describe in adding references to things you don't need to break conflicts... That's actually the real solution for these things in .NET Core. Many tiny dependencies, patches sometimes require you to reference things directly that are only used in a transitive fashion... It's not expected from full .NET framework projects, but as the lines blur, that is more and more a thing.

Not saying there's nothing we can do, but setting expectations - it may well be that the official answer is to add that reference to force a specific version from your project.

**Comment by mrahhal** [comment]
You're right, perhaps calling it a "workaround" is a bit misplaced, but _it really is a workaround_ in this particular case. Since the transitive dependencies are just 2 levels really (You depend on DI Abstractions which depends on a higher version of netstdlib than you do).

And much more importantly, referencing a single package should just work. To me, this "workaround" is a real solution only when the conflict is caused by depending on _multiple_ packages that each reference a conflicting version (wrt the host app), as opposed to a single lib causing it.

**Comment by tillig** [comment]
It doesn't really matter how many levels of dependencies it is if there are projects where this totally _isn't seen at all_. Changing any of our dependencies is a challenge because it fixes one thing but can break others. Like, it might fix _this specific case_ but break _all the currently working projects_.

Like I said, I'm not saying there won't be any action here, I just want to set expectations. We're definitely not ready to issue a new major release at this time.

**Comment by mrahhal** [comment]
My point was "Referencing a single package should just work". But I understand. Feel free to close or keep this issue.

**Comment by valeriob** [comment]
is this related to https://github.com/aspnet/Home/issues/2900 ?
How come this package exists ? https://www.nuget.org/packages/Alexinea.Autofac.Extensions.DependencyInjection/4.2.0

**Comment by tillig** [comment]
I've finally gotten a chance to look at the repro. I see it targets `net461`. There appears to be some odd build tooling challenges [when handling the `NetStandardImplicitPackageVersion`](https://docs.microsoft.com/en-us/dotnet/core/packages) when it comes to full .NET framework packages. For example, if I switch the repro to target `netcoreapp2.0` the warning goes away.

It also seems that current build tooling sets the implicit NETStandard.Library package version to 1.6.1 with no way to override that from the .csproj level. Odd and unfortunate.

We seem to have somewhat escaped this in core Autofac by having a specific `net45` target. It's unclear if that's the solution here. Given the desire to maintain compatibility with .NET Core 1.x apps and `netstandard1.1` interfaces, I'll have to look and see if there's a way to continue referencing `Microsoft.Extensions.DependencyInjection.Abstractions` 1.1.0 and solve the problem at the package level. Given build tooling has changed quite a bit since we originally published seven months ago, maybe it's as simple as rebuilding the package.

Let me mess with it. Stay tuned.

**Comment by valeriob** [comment]
awesome @tillig thanks for the info, but i'm familiar with it, i just do not know the tool diagnose those errors 😢 can you point me to the problem (the first issue links the output of the runtime exception and the deps.json file)
what made me suspicious is the existence of Aleaxinea.Autofac.Extensions.DependencyInjection package if you check (i contacted the author), the only difference is that it multitarget netstandard2.0;netstandard1.6;net461.
I think is a simptom of other ppl having the same issue.

**Comment by tillig** [comment]
@valeriob There's no "tool," you just need to Google. David Fowler posted [this StackOverflow answer](https://stackoverflow.com/questions/29007962/how-can-i-diagnose-missing-dependencies-or-other-loader-failures-in-dnx) that has tips. I would recommend pursuing that with them; that's well outside the purview of Autofac.

Again, can't speak to that package. It'd probably be better to truncate that portion of this issue here since it's just going to confuse things. Thanks.

**Comment by tillig** [comment]
Published [v4.2.1](https://www.nuget.org/packages/Autofac.Extensions.DependencyInjection/4.2.1) which should fix the issue. Updated build tooling, updated Autofac core reference to 4.0.1, and updated the implicit NETStandard.Library reference.
